File: src/com/train/algorithm/dynamicProgramming/implementInC++/LongestArithmeticSubsequenceofGivenDifference.cpp
//
// Created by 罗斯 on 24/11/2022.
//
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        unordered_map<int, int> dp;
        int res = 0;
        for (int a: arr) {
            res = max(res, dp[a] = dp[a - difference] + 1);
        }
        return res;
    }
};

File: src/com/train/algorithm/dynamicProgramming/implementInPython/LongestArithmeticSubsequenceOfGivenDifference.py
from collections import defaultdict


class Solution(object):
    def longestSubsequence(self, arr, difference):
        dp = {}
        max_len = 1
        for num in arr:
            if num - difference in dp:
                dp[num] = dp[num - difference] + 1
            else:
                dp[num] = 1
            max_len = max(max_len, dp[num])
        return max_len